### PR TITLE
Link hive account

### PR DESCRIPTION
**Pull request by sisygoboom**
- link hive account
- staging dockerfile uses projects ts config
- mock auth guard fixed incorrect assumption
- Improved hive verification functions

### CODE REVIEW

File: Dockerfile.staging [hunk 1]
```diff
@@ -4,5 +4,5 @@ COPY package*.json ./
 RUN apk add --no-cache git python3 make g++ ffmpeg
 RUN npm install && npm install typescript -g
 COPY . .
-RUN tsc
+RUN tsc -p tsconfig.json
```
Comment (Geo25rey) [inline Dockerfile.staging:7]: Isn't this the default path?

File: src/services/hive/hive.service.ts [hunk 2]
```diff
@@ -69,21 +77,20 @@ export class HiveService {
...
     if (!process.env.ACCOUNT_CREATOR) {
       throw new Error('Please set the ACCOUNT_CREATOR env var');
     }
     try {
-      return await this.#hiveRepository.createAccountWithAuthority(
+      await this.#linkedAccountsRepository.linkHiveAccount(sub, hiveUsername);
+      return await this.#hiveChainRepository.createAccountWithAuthority(
         hiveUsername,
         process.env.ACCOUNT_CREATOR,
       );
```
Comment (Geo25rey) [inline src/services/hive/hive.service.ts:94]: These operations need to be flipped in case that the account creation fails. We don't want to give anyone easy access to linking with some else's account

File: src/repositories/linked-accounts/linked-account.repository.ts [hunk 3]
```diff
@@ -23,8 +23,12 @@ export class LinkedAccountRepository {
   async findOneByUserIdAndAccountName(query: {
     account: LinkedAccount['account'];
     user_id: LinkedAccount['user_id'];
+    network: LinkedAccount['network'];
```
Comment (Geo25rey) [inline src/repositories/linked-accounts/linked-account.repository.ts:26]: Isn't network implied by `user_id` AKA `sub`?

File: src/services/auth/auth.controller.test.ts [hunk 4]
```diff
@@ -140,37 +140,43 @@ describe('AuthController', () => {
...
-        .then(response => {
+        .then(async response => {
           expect(response.body).toEqual({
             block_num: 1,
             expired: false,
             id: "id",
             trx_num: 10,
           });
+          
+          expect(await hiveService.isHiveAccountLinked('singleton/bob/did', hiveUsername)).toBe(true)
```
Comment (Geo25rey) [inline src/services/auth/auth.controller.test.ts:160]: The test function is already `async`. I think making a few variables and `await`ing the promises is better. I guess the same applies in many places actually. I'd recommend to do the refactor in a different PR.

File: Dockerfile.staging (hunk 1 above)
Comment (sisygoboom) [reply Dockerfile.staging:7]: This using default typescript settings unless a file is specified would explain why the build failed on staging but passed the CI typecheck

File: src/services/auth/auth.controller.test.ts (hunk 4 above)
Comment (sisygoboom) [reply src/services/auth/auth.controller.test.ts:160]: Maybe so; I see no harm in doing it this way for now. Maybe a task for after release

File: Dockerfile.staging (hunk 1 above)
Comment (Geo25rey) [reply Dockerfile.staging:7]: [The CI](https://github.com/spknetwork/acela-core/blob/main/.github/workflows/node.js.yml#L29-L32) looks a lot different than the Dockerfile.

Also, maybe it's the `npm install typescript -g` this would install the latest version of typescript while the CI does not.

File: src/services/auth/auth.controller.test.ts (hunk 4 above)
Comment (Geo25rey) [reply src/services/auth/auth.controller.test.ts:160]: I think this is fine for this PR, but next time you update a test like this, I think it is worth taking the extra time to clean it up a little.

File: src/repositories/linked-accounts/linked-account.repository.ts [hunk 5]
```diff
@@ -23,12 +23,11 @@ export class LinkedAccountRepository {
   async findOneByUserIdAndAccountName(query: {
     account: LinkedAccount['account'];
     user_id: LinkedAccount['user_id'];
-    network: LinkedAccount['network'];
   }) {
     return await this.linkedAccountModel.findOne({
       ...query,
       status: 'verified',
-    } satisfies LinkedAccount);
+    } satisfies Partial<LinkedAccount>);
```
Comment (Geo25rey) [inline src/repositories/linked-accounts/linked-account.repository.ts:30]: Doesn't this already have type checking so `satisfies` wouldn't be required?

Comment (sisygoboom) [reply src/repositories/linked-accounts/linked-account.repository.ts:30]: No, mongoose has terrible type-checking, this could fail at runtime due to the spread operator (possibly?) - I can't remember exactly but it was needed for some reason.

File: Dockerfile.staging (hunk 1 above)
Comment (sisygoboom) [reply Dockerfile.staging:7]: if you look at the package json, `npm run tsc:noemit` is actually `npm run tsc -p tsconfig.json --noEmit` under the hood.

Comment (Geo25rey) [reply Dockerfile.staging:7]: My original point was isn't running `tsc` equivalent to `tsc -p tsconfig.json` since "tsconfig.json" is the default path?

File: src/repositories/linked-accounts/linked-account.repository.ts (hunk 5 above)
Comment (Geo25rey) [reply src/repositories/linked-accounts/linked-account.repository.ts:30]: I'm not familiar enough with mongoose to know if it is broken or not. I think it is worth double checking. Can you test this?

Comment (sisygoboom) [reply src/repositories/linked-accounts/linked-account.repository.ts:30]: image

File: Dockerfile.staging (hunk 1 above)
Comment (sisygoboom) [reply Dockerfile.staging:7]: The non-prod docker file uses `RUN npm run tsc:noemit` and ran without any type errors

File: src/repositories/linked-accounts/linked-account.repository.ts (hunk 5 above)
Comment (Geo25rey) [reply src/repositories/linked-accounts/linked-account.repository.ts:30]: Weird. I see. As long as there is nothing missed with added the `LinkedAccount` type to `linkedAccountModel`, then I think this is fine.

File: Dockerfile.staging (hunk 1 above)
Comment (Geo25rey) [reply Dockerfile.staging:7]: I think that part is a red-haring, since `tsc` and `tsc -p tsconfig.json` should be equivalent. Please consider my previous point:
> Also, maybe it's the npm install typescript -g this would install the latest version of typescript while the CI does not.

So, the different versions of typescript can cause different type checking behaviors.

With these 2 points, I think this change does not fix the issue, but removing `npm install typescript -g` from the Dockerfile should.

Comment (Geo25rey) [reply Dockerfile.staging:7]: Looks like you have included the requested change in #30, so I think we can mark this as resolved.

**Review by Geo25rey — APPROVED** [review]
